### ISSUE TITLE
serious reproducibility issue with PHMLinear layers 

### ISSUE THREAD

**Issue by dorooddorood606** [bayer-science-for-a-better-life/phc-gnn#3]
Hi there, 

I am using  PHMLinear here [1], then I set all the seeds (torch/torch cuda/numpy/random) once before making a model, and then I use this layer inside the model, I am getting each time a different results, could you guide me please how I can fix this issue with this layer?

thank you 

[1] https://github.com/bayer-science-for-a-better-life/phc-gnn/blob/f4f04d1a0350058599380b0d86426f4d7b7eb412/phc/hypercomplex/layers.py#L113-L191

**Comment by tuanle618** [comment]
Hi @dorooddorood606,

I tested the `PHMLinear` layer, and when I fix the random-seed, as defined in our function (https://github.com/bayer-science-for-a-better-life/phc-gnn/blob/master/benchmarks/utils.py#L22-L26), I obtain the same weights `self.W`, `self.bias` and `self.phm_rule` for the phm-linear layers as well as the output vector `y`, when I compute a forward pass, see below:

```python
from benchmarks.utils import set_seed_all
from phc.hypercomplex.layers import PHMLinear
import torch

set_seed_all(seed=42)
x = torch.randn(512, 128)

set_seed_all(seed=43)
phm_lin1 = PHMLinear(in_features=128 // 2, out_features=128 // 2, phm_dim=2, w_init="phm", c_init="standard")
weights1 = [m.data for m in phm_lin1.parameters() if m.requires_grad]

set_seed_all(seed=43)
phm_lin2 = PHMLinear(in_features=128 // 2, out_features=128 // 2, phm_dim=2, w_init="phm", c_init="standard")
weights2 = [m.data for m in phm_lin2.parameters() if m.requires_grad]

bools = []
for w_1, w_2 in zip(weights1, weights2):
    bools.append(torch.allclose(w_1, w_2))

print(bools)
# [True, True, True, True, True, True]

y1 = phm_lin1(x)
y2 = phm_lin2(x)

torch.allclose(y1, y2)
# True
```

How did you set the seed? You are talking about a **model**. It might be worth checking, if your other modules are initialised differently.

Best regards,
Tuan

**Comment by dorooddorood606** [comment]
Thanks a lot, I realized the issue was on my side
